federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+		}
+
+		replicas := replicaNums{min: 0, max: 0}
+		scheduleStatus[cluster] = &replicas
+		if obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MinReplicas != nil {
+			existingTotal.min += *obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MinReplicas
+			replicas.min += *obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MinReplicas
+		}
+		existingTotal.max += obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MaxReplicas
+		replicas.max += obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MaxReplicas
+	}
+
+	return lists, existingTotal, scheduleStatus
+}
+
+// reduceMinReplicas reduces the min replicas from existing clusters.
+// At the end of the function remainingReplicas.min should be 0 and the MinList
+// and the scheduledReplicas properly updated in place.
+func reduceMinReplicas(remainingMin int32, availableMinList sets.String, scheduled map[string]*replicaNums) int32 {
+	if remainingMin < 0 {
+		// first we try reducing from those clusters which already offer min
+		if availableMinList.Len() > 0 {
+			for _, cluster := range availableMinList.List() {
+				replicas := scheduled[cluster]
+				if (replicas.min - 1) != 0 {
+					replicas.min--
+					availableMinList.Delete(cluster)
+					remainingMin++
+					if remainingMin == 0 {
+						break
+					}
+				}
+			}
+		}
+	}
+
+	// If we could not get needed replicas from already offered min above
+	// we abruptly start removing replicas from some/all clusters.
+	// Here we might make some min to 0 signalling that this hpa might be a
+	// candidate to be removed from this cluster altogether.
+	for remainingMin < 0 {
+		for _, replicas := range scheduled {
+			if replicas != nil &&
+				!((replicas.min - 1) < 0) {
+				replicas.min--
+				remainingMin++
+				if remainingMin == 0 {
+					break
+				}
+			}
+		}
+	}
+
+	return remainingMin
+}
+
+func reduceMaxReplicas(remainingMax int32, availableMaxList sets.String, scheduled map[string]*replicaNums) int32 {
+	if remainingMax < 0 {
+		// first we try reducing from those clusters which already offer max
+		if availableMaxList.Len() > 0 {
+			for _, cluster := range availableMaxList.List() {
+				replicas := scheduled[cluster]
+				if replicas != nil && !((replicas.max - replicas.min) < 0) {
+					replicas.max--
+					availableMaxList.Delete(cluster)
+					remainingMax++
+					if remainingMax == 0 {
+						break
+					}
+				}
+			}
+		}
+	}
+	// If we could not get needed replicas to reduce from already offered
+	// max above we abruptly start removing replicas from some/all clusters.
+	// Here we might make some max and min to 0, signalling that this hpa be
+	// removed from this cluster altogether
+	for remainingMax < 0 {
+		for _, replicas := range scheduled {
+			if replicas != nil &&
+				!((replicas.max - replicas.min) < 0) {
+				replicas.max--
+				remainingMax++
+				if remainingMax == 0 {
+					break
+				}
+			}
+		}
+	}
+
+	return remainingMax
+}
+
+func distributeMaxReplicas(toDistributeMax int32, lists hpaLists, rdc replicaNums,
```

Comment:
Updated with comment for the function details.
About rcd, yes its replicaDistributionCount, I had a comment to explain the name, at its declaration. I found the full name too long to be used everywhere.